// @progress/kendo-vue-dropdowns/dist/npm/common/ClearButton.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ClearButton = void 0; // @ts-ignore

var Vue = require("vue");

var allVue = Vue;
var gh = allVue.h;
var inject = allVue.inject;

var kendo_vue_intl_1 = require("@progress/kendo-vue-intl");

var messages_1 = require("../messages");
/**
 * Represents the default `ClearButton` component.
 */


var ClearButton = {
  name: 'clear-button',
  inject: {
    kendoLocalizationService: {
      default: null
    }
  },
  // @ts-ignore
  setup: !gh ? undefined : function () {
    var v3 = !!gh;
    var kendoLocalizationService = inject('kendoLocalizationService', {});
    return {
      v3: v3,
      kendoLocalizationService: kendoLocalizationService
    };
  },
  methods: {
    onMouseDown: function onMouseDown(e) {
      e.preventDefault();
    },
    onClickHandler: function onClickHandler(e) {
      this.$emit('clearclick', e);
    }
  },
  // @ts-ignore
  render: function render(createElement) {
    var h = gh || createElement;
    var title = kendo_vue_intl_1.provideLocalizationService(this).toLanguageString(messages_1.clear, messages_1.messages[messages_1.clear]);
    return h("span", {
      "class": "k-icon k-clear-value k-i-close",
      role: "button",
      attrs: this.v3 ? undefined : {
        role: "button",
        tabIndex: -1,
        title: title
      },
      onClick: this.onClickHandler,
      on: this.v3 ? undefined : {
        "click": this.onClickHandler,
        "mousedown": this.onMouseDown
      },
      onMousedown: this.onMouseDown,
      tabIndex: -1,
      title: title,
      key: "clearbutton"
    });
  }
};
exports.ClearButton = ClearButton;
