// @progress/kendo-vue-dropdowns/dist/npm/common/ListDefaultItem.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ListDefaultItem = void 0; // @ts-ignore

var Vue = require("vue");

var allVue = Vue;
var gh = allVue.h;

var kendo_vue_common_1 = require("@progress/kendo-vue-common");

var utils_1 = require("./utils");
/**
 * Represents the default `ListDefaultItem` component.
 */


var ListDefaultItem = {
  name: 'list-default-item',
  props: {
    defaultItem: [Object, String],
    textField: String,
    selected: Boolean
  },
  // @ts-ignore
  setup: !gh ? undefined : function () {
    var v3 = !!gh;
    return {
      v3: v3
    };
  },
  methods: {
    onMouseDown: function onMouseDown(e) {
      e.preventDefault();
    },
    onClick: function onClick(e) {
      this.$emit('defaultitemclick', e);
    }
  },
  // @ts-ignore
  render: function render(createElement) {
    var h = gh || createElement;
    var _a = this.$props,
        selected = _a.selected,
        defaultItem = _a.defaultItem,
        textField = _a.textField;
    return h("div", {
      onClick: this.onClick,
      on: this.v3 ? undefined : {
        "click": this.onClick,
        "mousedown": this.onMouseDown
      },
      onMousedown: this.onMouseDown,
      "class": kendo_vue_common_1.classNames('k-list-optionlabel', {
        'k-state-selected': selected
      })
    }, [utils_1.getItemValue(defaultItem, textField) || '']);
  }
};
exports.ListDefaultItem = ListDefaultItem;
