// @progress/kendo-treeview-vue-wrapper/dist/npm/KendoTreeView/Animation.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.default = {
    props: {
        animationCollapse: {
            type: Boolean,
            default: undefined,
            kComposite: 'animation.collapse'
        },
        animationCollapseDuration: {
            type: Number,
            default: undefined,
            kComposite: 'animation.collapse.duration'
        },
        animationCollapseEffects: {
            type: String,
            default: undefined,
            kComposite: 'animation.collapse.effects'
        },
        animationExpand: {
            type: Boolean,
            default: undefined,
            kComposite: 'animation.expand'
        },
        animationExpandDuration: {
            type: Number,
            default: undefined,
            kComposite: 'animation.expand.duration'
        },
        animationExpandEffects: {
            type: String,
            default: undefined,
            kComposite: 'animation.expand.effects'
        }
    }
};
