// @progress/kendo-vue-inputs/dist/npm/messages/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.messages = exports.checkboxOptionalText = exports.checkboxValidation = exports.colorGradientHex = exports.colorGradientA = exports.colorGradientB = exports.colorGradientG = exports.colorGradientR = exports.sliderDragTitle = exports.sliderDecreaseValue = exports.sliderIncreaseValue = exports.numericDecreaseValue = exports.numericIncreaseValue = void 0;
/**
 * @hidden
 */
exports.numericIncreaseValue = 'numerictextbox.increment';
/**
 * @hidden
 */
exports.numericDecreaseValue = 'numerictextbox.decrement';
/**
 * @hidden
 */
exports.sliderIncreaseValue = 'slider.increment';
/**
 * @hidden
 */
exports.sliderDecreaseValue = 'slider.decrement';
/**
 * @hidden
 */
exports.sliderDragTitle = 'slider.dragTitle';
/**
 * @hidden
 */
exports.colorGradientR = 'colorGradient.r';
/**
 * @hidden
 */
exports.colorGradientG = 'colorGradient.g';
/**
 * @hidden
 */
exports.colorGradientB = 'colorGradient.b';
/**
 * @hidden
 */
exports.colorGradientA = 'colorGradient.a';
/**
 * @hidden
 */
exports.colorGradientHex = 'colorGradient.hex';
/**
 * @hidden
 */
exports.checkboxValidation = 'checkbox.validation';
/**
 * @hidden
 */
exports.checkboxOptionalText = 'checkbox.optionalText';
/**
 * @hidden
 */
exports.messages = (_a = {},
    _a[exports.numericIncreaseValue] = 'Increase value',
    _a[exports.numericDecreaseValue] = 'Decrease value',
    _a[exports.sliderIncreaseValue] = 'Increase',
    _a[exports.sliderDecreaseValue] = 'Decrease',
    _a[exports.sliderDragTitle] = 'Drag',
    _a[exports.colorGradientR] = 'r',
    _a[exports.colorGradientG] = 'g',
    _a[exports.colorGradientB] = 'b',
    _a[exports.colorGradientA] = 'a',
    _a[exports.colorGradientHex] = 'hex',
    _a);
