// @progress/kendo-treelist-vue-wrapper/dist/npm/KendoTreeList/KendoTreeListMixin.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _kendoBaseComponentsVueWrapper = require('@progress/kendo-base-components-vue-wrapper');

var _TreeListProps = require('./TreeListProps');

var _TreeListProps2 = _interopRequireDefault(_TreeListProps);

var _KendoTreeListColumn = require('../KendoTreeListColumn');

var _KendoTreeListColumn2 = _interopRequireDefault(_KendoTreeListColumn);

var _ColumnMenu = require('./ColumnMenu');

var _ColumnMenu2 = _interopRequireDefault(_ColumnMenu);

var _Editable = require('./Editable');

var _Editable2 = _interopRequireDefault(_Editable);

var _Excel = require('./Excel');

var _Excel2 = _interopRequireDefault(_Excel);

var _Filterable = require('./Filterable');

var _Filterable2 = _interopRequireDefault(_Filterable);

var _Messages = require('./Messages');

var _Messages2 = _interopRequireDefault(_Messages);

var _Pageable = require('./Pageable');

var _Pageable2 = _interopRequireDefault(_Pageable);

var _Pdf = require('./Pdf');

var _Pdf2 = _interopRequireDefault(_Pdf);

var _Sortable = require('./Sortable');

var _Sortable2 = _interopRequireDefault(_Sortable);

var _Toolbar = require('./Toolbar');

var _Toolbar2 = _interopRequireDefault(_Toolbar);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = {
    props: _TreeListProps2.default,
    emits: {
        'kendowidgetready': null
    },
    mixins: [_kendoBaseComponentsVueWrapper.KendoBaseComponent, _kendoBaseComponentsVueWrapper.KendoBaseDatasourceComponent, _ColumnMenu2.default, _Editable2.default, _Excel2.default, _Filterable2.default, _Messages2.default, _Pageable2.default, _Pdf2.default, _Sortable2.default, _Toolbar2.default],
    mounted() {
        this._resolveChildren();

        kendo.jQuery(this.$el).kendoTreeList(this.widgetOptions);

        if (this.v3) {
            this.ready();
            this.$emit('kendowidgetready', this.kendoWidget());
        } else {
            this.$emit('kendowidgetready', this.kendoWidget());
        }
    },
    methods: {
        kendoWidget() {
            return kendo.jQuery(this.$el).getKendoTreeList();
        },
        _resolveChildren() {
            this.resolveChildren('columns', _KendoTreeListColumn2.default.name);
        }
    }
};
