// @progress/kendo-vue-dateinputs/dist/npm/calendar/utils/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.extractActiveRange = exports.extractFocusedDate = exports.extractRangeFromValue = exports.extractMultipleFromValue = exports.extractDateFromValue = exports.calculateValue = void 0;
var utils_1 = require("../../utils");
var models_1 = require("../models");
var kendo_date_math_1 = require("@progress/kendo-date-math");
var calculateValue = function (min, max, stateValue, propValue) {
    return propValue !== undefined
        ? propValue !== null && utils_1.isInRange(propValue, min, max)
            ? propValue
            : null
        : stateValue !== null && utils_1.isInRange(stateValue, min, max)
            ? stateValue
            : null;
};
exports.calculateValue = calculateValue;
var extractDateFromValue = function (min, max, value) {
    return value instanceof Date && !Array.isArray(value) && utils_1.isInRange(kendo_date_math_1.getDate(value), min, max)
        ? kendo_date_math_1.getDate(value)
        : null;
};
exports.extractDateFromValue = extractDateFromValue;
var extractMultipleFromValue = function (min, max, value) {
    return Array.isArray(value)
        ? value.filter(function (date) { return utils_1.isInRange(date, min, max); }).map(function (date) { return kendo_date_math_1.getDate(date); })
        : null;
};
exports.extractMultipleFromValue = extractMultipleFromValue;
var extractRangeFromValue = function (value) {
    return typeof value === 'object' && !(value instanceof Date) && value !== null && !Array.isArray(value)
        ? value
        : models_1.EMPTY_SELECTIONRANGE;
};
exports.extractRangeFromValue = extractRangeFromValue;
var extractFocusedDate = function (single, multiple, range) {
    return single || (multiple && multiple[0]) || (range && range.start);
};
exports.extractFocusedDate = extractFocusedDate;
var extractActiveRange = function (range, single) {
    return range.start === null && single === null
        ? 'start'
        : range.end === null
            ? 'end'
            : 'start';
};
exports.extractActiveRange = extractActiveRange;
