// @progress/kendo-vue-dateinputs/dist/npm/timepicker/TimePart.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TimePart = exports.Direction = void 0; // @ts-ignore

var Vue = require("vue");

var allVue = Vue;
var gh = allVue.h;
var inject = allVue.inject;

var kendo_date_math_1 = require("@progress/kendo-date-math");

var kendo_vue_common_1 = require("@progress/kendo-vue-common");

var kendo_vue_intl_1 = require("@progress/kendo-vue-intl");

var messages_1 = require("../messages");

var TimeList_1 = require("./TimeList");

var utils_1 = require("../utils");

var TimePart_1 = require("./models/TimePart");

var utils_2 = require("./utils");

var formatRegExp = new RegExp(TimePart_1.TIME_PART.hour + "|" + TimePart_1.TIME_PART.minute + "|" + TimePart_1.TIME_PART.second + "|" + TimePart_1.TIME_PART.dayperiod + "|literal");
/**
 * @hidden
 */

var Direction;

(function (Direction) {
  Direction[Direction["Left"] = 0] = "Left";
  Direction[Direction["Right"] = 1] = "Right";
})(Direction = exports.Direction || (exports.Direction = {}));

var TimePart = {
  name: 'KendoTimePart',
  props: {
    cancelButton: {
      type: Boolean,
      default: true
    },
    disabled: {
      type: Boolean,
      default: false
    },
    format: {
      type: String,
      default: function _default() {
        return 'hh:mm a';
      }
    },
    max: {
      type: Date,
      default: function _default() {
        return utils_1.MAX_TIME;
      }
    },
    min: {
      type: Date,
      default: function _default() {
        return utils_1.MIN_TIME;
      }
    },
    nowButton: {
      type: Boolean,
      default: true
    },
    steps: {
      type: Object,
      default: function _default() {
        return {};
      }
    },
    smoothScroll: {
      type: Boolean,
      default: true
    },
    tabIndex: Number,
    value: {
      type: Date,
      default: function _default() {
        return null;
      }
    }
  },
  // @ts-ignore
  emits: {
    'change': null,
    'focus': null,
    'blur': null
  },
  created: function created() {
    this.timeLists = [];
    this.snapTime = utils_2.snapTime(utils_2.generateSnappers(this.$props.steps, this.$props.min));
    this.activeListIndex = -1;
    this.hasActiveButton = this.hasActiveButton.bind(this);
  },
  inject: {
    kendoIntlService: {
      default: null
    },
    kendoLocalizationService: {
      default: null
    }
  },
  data: function data() {
    return {
      activeListIndex: null
    };
  },
  computed: {
    element: function element() {
      return this._element;
    },
    computedValue: function computedValue() {
      return utils_2.timeInRange(this.snapTime(kendo_date_math_1.cloneDate(this.$props.value || utils_1.MIDNIGHT_DATE)), this.computedMin, this.computedMax);
    },
    intl: function intl() {
      return kendo_vue_intl_1.provideIntlService(this);
    },
    computedMin: function computedMin() {
      return this.snapTime(this.$props.min);
    },
    computedMax: function computedMax() {
      return this.snapTime(this.$props.max);
    }
  },
  mounted: function mounted() {
    var _this = this;

    this._nowButton = this.$refs.nowButton;
    this.dateFormatParts.forEach(function (item, idx) {
      if (item.type !== 'literal') {
        _this.timeLists.push(_this.$refs['timeList' + idx]);
      }
    });
  },
  // @ts-ignore
  setup: !gh ? undefined : function () {
    var v3 = !!gh;
    var kendoIntlService = inject('kendoIntlService', {});
    var kendoLocalizationService = inject('kendoLocalizationService', {});
    return {
      v3: v3,
      kendoIntlService: kendoIntlService,
      kendoLocalizationService: kendoLocalizationService
    };
  },
  // @ts-ignore
  render: function render(createElement) {
    var h = gh || createElement;
    var _a = this.$props,
        format = _a.format,
        smoothScroll = _a.smoothScroll,
        disabled = _a.disabled;
    this.snapTime = utils_2.snapTime(utils_2.generateSnappers(this.$props.steps, this.computedMin));
    this.dateFormatParts = this.intl.splitDateFormat(format).filter(this.timeFormatFilter);
    var rootClassName = kendo_vue_common_1.classNames({
      'k-state-disabled': disabled
    }, 'k-time-part');
    var localizationService = kendo_vue_intl_1.provideLocalizationService(this);
    var selectNowMessage = localizationService.toLanguageString(messages_1.selectNow, messages_1.messages[messages_1.selectNow]);
    return h("div", {
      "class": rootClassName
    }, [h("div", {
      "class": "k-time-header"
    }, [h("span", {
      "class": "k-title"
    }, [this.intl.formatDate(this.computedValue, this.dateFormatParts.reduce(this.timeFormatReducer, ''))]), this.showNowButton() && h("button", {
      ref: 'nowButton',
      "class": "k-button k-flat k-time-now",
      title: selectNowMessage,
      attrs: this.v3 ? undefined : {
        title: selectNowMessage,
        "aria-label": selectNowMessage,
        tabIndex: disabled ? -1 : 0
      },
      "aria-label": selectNowMessage,
      onClick: this.onNowClick,
      on: this.v3 ? undefined : {
        "click": this.onNowClick
      },
      tabIndex: disabled ? -1 : 0
    }, [localizationService.toLanguageString(messages_1.now, messages_1.messages[messages_1.now])])]), h("div", {
      "class": "k-time-list-container",
      onKeydown: this.handleKeyDown,
      on: this.v3 ? undefined : {
        "keydown": this.handleKeyDown
      }
    }, [h("span", {
      "class": "k-time-highlight"
    }), this.dateFormatParts.map(function (part, idx) {
      var _this = this;

      return part.type !== 'literal' ? h("div", {
        key: idx,
        "class": kendo_vue_common_1.classNames('k-time-list-wrapper', {
          'k-state-focused': idx === this.activeListIndex
        }),
        role: "presentation",
        attrs: this.v3 ? undefined : {
          role: "presentation",
          tabIndex: -1
        },
        tabIndex: -1
      }, [h("span", {
        "class": "k-title",
        onMousedown: function onMousedown(e) {
          e.preventDefault();
        },
        on: this.v3 ? undefined : {
          "mousedown": function onMousedown(e) {
            e.preventDefault();
          }
        }
      }, [this.intl.dateFieldName(part)]), // @ts-ignore  function children
      h(TimeList_1.TimeList, {
        min: this.computedMin,
        attrs: this.v3 ? undefined : {
          min: this.computedMin,
          max: this.computedMax,
          boundRange: this.$props.boundRange,
          part: part,
          step: part.type ? this.$props.steps[part.type] : 1,
          smoothScroll: smoothScroll,
          id: idx,
          value: this.computedValue,
          disabled: disabled
        },
        max: this.computedMax,
        boundRange: this.$props.boundRange,
        part: part,
        step: part.type ? this.$props.steps[part.type] : 1,
        smoothScroll: smoothScroll,
        ref: 'timeList' + idx,
        id: idx,
        onFocus: function onFocus(event) {
          _this.handleListFocus(event, idx);
        },
        on: this.v3 ? undefined : {
          "focus": function onFocus(event) {
            _this.handleListFocus(event, idx);
          },
          "blur": this.handleListBlur,
          "change": this.handleChange
        },
        onBlur: this.handleListBlur,
        onChange: this.handleChange,
        value: this.computedValue,
        disabled: disabled
      })]) : h("div", {
        key: idx,
        "class": "k-time-separator"
      }, [part.pattern]);
    }, this)])]);
  },
  methods: {
    onNowClick: function onNowClick(event) {
      this.$emit('nowclick', event);
    },
    focus: function focus(args) {
      var _this = this;

      this.$nextTick(function () {
        var timeList = _this.timeLists[0];

        if (!_this.hasActiveButton() && timeList && timeList.$el) {
          timeList.focus(args);
        }
      });
    },
    timeFormatReducer: function timeFormatReducer(acc, current) {
      return acc + current.pattern;
    },
    timeFormatFilter: function timeFormatFilter(part, index, all) {
      var prevItem = index >= 1 && all[index - 1];

      if (!prevItem) {
        return formatRegExp.test(part.type || '');
      }

      if (prevItem && part.type === 'literal') {
        return formatRegExp.test(prevItem.type || '');
      }

      return formatRegExp.test(part.type || '');
    },
    hasActiveButton: function hasActiveButton() {
      return document.activeElement === this._nowButton;
    },
    focusList: function focusList(dir) {
      if (!this.timeLists.length) {
        return;
      }

      this.timeLists.reduce(this.listReducer, []).map(function (state) {
        return dir === Direction.Right ? state.next : state.prev;
      }).map(function (list) {
        return list && list.$el && list.$el.focus({
          preventScroll: true
        });
      });
    },
    listReducer: function listReducer(state, list, idx, all) {
      if (state.length || list.$props.id !== this.activeListIndex) {
        return state;
      }

      return [{
        next: all[idx + 1] || list,
        prev: all[idx - 1] || list
      }];
    },
    showNowButton: function showNowButton() {
      return !this.hasSteps() && this.$props.nowButton && utils_2.isInTimeRange(utils_2.getNow(), this.computedMin, this.computedMax);
    },
    hasSteps: function hasSteps() {
      var _this = this;

      var keys = Object.keys(this.$props.steps);
      return keys.length !== keys.reduce(function (acc, k) {
        return acc + _this.$props.steps[k];
      }, 0);
    },
    handleKeyDown: function handleKeyDown(event) {
      var keyCode = event.keyCode;

      switch (keyCode) {
        case kendo_vue_common_1.Keys.left:
          event.preventDefault();
          this.focusList(Direction.Left);
          return;

        case kendo_vue_common_1.Keys.right:
          event.preventDefault();
          this.focusList(Direction.Right);
          return;

        default:
          return;
      }
    },
    handleListBlur: function handleListBlur(event) {
      this.$emit('blur', event);
    },
    handleListFocus: function handleListFocus(event, idx) {
      this.$emit('focus', event);
      this.activeListIndex = idx;
    },
    handleChange: function handleChange(candidate) {
      this.$emit('change', candidate);
    }
  }
};
exports.TimePart = TimePart;
