// @progress/kendo-vue-intl/dist/npm/Intl/IntlProvider.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.IntlProvider = void 0; // @ts-ignore

var aVue = require("vue");

var allVue = aVue;
var gh = allVue.h;
var ref = allVue.ref;
var provide = allVue.provide;

var IntlService_1 = require("./IntlService");

var kendo_vue_common_1 = require("@progress/kendo-vue-common");

var kendo_licensing_1 = require("@progress/kendo-licensing");

var package_metadata_1 = require("../package-metadata");
/**
 * A Vue component which provides an internationalization service.
 * Expects a locale string as a property of the component.
 */


var IntlProvider = {
  props: {
    locale: String
  },
  data: function data() {
    return this.getChildContext();
  },
  // @ts-ignore
  setup: function setup(props) {
    var v3 = !!gh;
    var intlService = ref(new IntlService_1.IntlService(props.locale));
    provide('kendoIntlService', intlService);
    return {
      v3: v3
    };
  },
  watch: {
    locale: function locale(newLocale) {
      this.$data.kendoIntlService.locale = newLocale;
    }
  },
  provide: function provide() {
    return {
      kendoIntlService: this.$data.kendoIntlService
    };
  },
  methods: {
    /**
     * Returns an internationalization service.
     * The method is suitable for overriding when you
     * implement custom internationalization behavior.
     */
    getIntlService: function getIntlService() {
      return new IntlService_1.IntlService(this.$props.locale);
    },

    /**
     * @hidden
     */
    getChildContext: function getChildContext() {
      return {
        kendoIntlService: this.getIntlService()
      };
    }
  },
  created: function created() {
    kendo_licensing_1.validatePackage(package_metadata_1.packageMetadata);
  },

  /**
   * @hidden
   */
  // @ts-ignore
  render: function render(createElement) {
    // @ts-ignore
    var h = gh || createElement;
    var defaultSlot = kendo_vue_common_1.getDefaultSlots(this);
    return h("div", [defaultSlot]);
  }
};
exports.IntlProvider = IntlProvider;
