// @progress/kendo-vue-animation/src/Expand.tsx
// @ts-ignore
import * as Vue from 'vue';
const allVue = Vue as any;
const gh = allVue.h;
import { RecordPropsDefinition, ComponentOptions } from 'vue/types/options';
type DefaultData<V> =  object | ((this: V) => {});
type DefaultMethods<V> =  { [key: string]: (this: V, ...args: any[]) => any };
import { Animation } from './Animation';
import { AnimationInterface } from './AnimationInterface';
import { VNode } from 'vue';
import { getDefaultSlots } from '@progress/kendo-vue-common';

/**
 * Specifies the direction of the Expand Animation ([see example]({% slug direction_animation_native %}#toc-expand)).
 *
 * The supported directions are:
 * * (Default) `vertical`&mdash;Expands the content from center to top and bottom, and vice-versa.
 * * `horizontal`&mdash;Expands the content from center to left and right, and vice-versa.
 */
export type ExpandDirection = 'horizontal' | 'vertical';

// tslint:disable:max-line-length
/**
 * Represent the props of the [KendoVue Expand Animation component]({% slug animationtypes_animation_native %}#toc-expand).
 *
 * @example
 * ```html-preview
 * 
 * ```
 */
export interface ExpandProps extends AnimationInterface {
    /**
     * After the element reached its exit state, it is no longer available in the DOM. If a DOM operation is required, access it trough the `childFactory` function.
     */
    childFactory?: any;
    /**
     * Specifies the CSS class names which are set to the Animation.
     */
    className?: string;
    /**
     * Specifies the direction of the Expand Animation. Defaults to `vertical`.
     */
    direction?: String;
    /**
     * Specifies the node type of the parent Animation. Defaults to `div`.
     */
    component?: string;
    /**
     * Specifies the tag.
     */
    tag?: string;
    /**
     * Specifies the id of the Animation.
     */
    id?: string;
    /**
     * Specifies the id of the Animation.
     */
    key?: string;
}

/**
 * Represents the default `Expand` component.
 */
let Expand: ComponentOptions<Vue, DefaultData<{}>, 
DefaultMethods<{}>, {}, RecordPropsDefinition<ExpandProps>> = {
    props: {
        appear: {
            type: Boolean,
            default: false
        },
        enter: {
            type: Boolean,
            default: true
        },
        exit: {
            type: Boolean,
            default: true
        },
        transitionEnterDuration: {
            type: Number,
            default: 300
        },
        transitionExitDuration: {
            type: Number,
            default: 300
        },
        childFactory: Object,
        className: String,
        direction: {
            type: String,
            default: 'vertical'
        },
        tag: String,
        id: String
    },
    // @ts-ignore
    setup: (!gh) ? undefined : function() {
        const v3 = !!gh;
        return {
            v3
        };
    },
    // @ts-ignore
    render(this: Vue, createElement: any): VNode {
        const h = gh || createElement;
        const defaultSlot = getDefaultSlots(this);
        const {
            direction,
            ...other
        } = this.$props;
        const transitionName = `expand-${this.$props.direction}`;

        return (
             // @ts-ignore function children
            <Animation 
                transitionName = {transitionName}
                appear = { this.$props.appear }
                enter = { this.$props.enter }
                exit = { this.$props.exit }
                transitionEnterDuration = { this.$props.transitionEnterDuration }
                transitionExitDuration = { this.$props.transitionExitDuration }
                >
                {defaultSlot}
            </Animation>
        );
    }
};

export { Expand };
